Python/chapter03/p02_stack_min/miguelHx.py

```diff
@@ -0,0 +1,248 @@
+"""
+Python version 3.7.0
+3.2 - Stack Min
+How would you design a stack which, in addition to
+push and pop, has a function min which returns the minimum element?
+Push, pop and min should all operate in O(1) time.
+"""
+
+import copy
+import unittest
+import sys
+
+from abc import abstractmethod
+from dataclasses import dataclass
+from typing import Generic, TypeVar
+from typing import List, Optional, Iterator
+
+T = TypeVar('T')
+C = TypeVar('C', bound='Comparable')
+
+class Comparable(Generic[T, C]):
+    @abstractmethod
+    def __lt__(self, other: T):
+        pass
+
+@dataclass
+class StackNode(Generic[T, C]):
+    data: T
+    next: 'Optional[StackNode[T, C]]'
+    running_min: T
+
+class MyStack(object):
+    """Stack data structure implementation.
+    Uses LIFO (last-in first-out) ordering.
+    The most recent item added to the stack is
+    the first removed.  Traversal is top to bottom.
+    """
+
+    def __init__(self):
+        self.top: Optional[StackNode[T]] = None # top is a pointer to StackNode object
+        self._size: int = 0
+        self.index: int = -1 # for iterator use
+        self.current_node: Optional[StackNode[T]] = self.top # also for iterator use
```

Comment:
optional: the most idiomatic way to express this is to have an additional class (maybe even embedded in this class as an inner class) that holds these members instead of the main class. this allows for multiple concurrent iterators:

https://anandology.com/python-practice-book/iterators.html#the-iteration-protocol

Reply:
Implemented here:  472f8f499950d068d2869b5ea4574cddeb7eb77d